[#4] 유저 비밀번호 변경기능 구현

msugo1 opened the pull request:
## 📖 구현 항목
### 🔖 비밀번호 변경 기능관련 로직 구현
- 회원의 현재 비밀번호를 먼저 확인받고 다음의 로직으로 넘어갑니다.
- 입력한 `현재 비밀번호가 일치하지 않으면` 예외를 발생시킵니다.
- `새로 입력한 비밀번호`가 `현재 비밀번호`와 `일치`하면 예외를 발생시킵니다.
- `새로 입력한 비밀번호`와 입력값을 `다시 확인하기 위한 값`이 `불일치` 하면 예외를 발생시킵니다.
- 위의 검증과정을 모두 통과하면 `DAO`를 통해 DB에서 현재 회원의 비밀번호를 업데이트 합니다.

### 🔖 위의 로직을 테스트하기 위한 테스트 작성

Commit:
[#4] 유저 비밀번호 변경 기능 구현

* UserPasswordUpdateRequest (new)
- 현재 비밀번호, 새로운 비밀번화, 새로운 비밀번호 확인 값을 담아 전달할
  객체를 생성하기 위한 클래스 작성

* SessionLoginService (update)
- 중복제거를 위한 메소드 추출: 세션에서 현재 유저의 ID를 얻기 위한
  getCurrentId() 메소드 추가

* CurrentUserArgumentResolver (new)
- 현재 로그인 중인 유저의 정보를 컨트롤러의 파라미터로 보낼 때  중복을 최대한으로 줄이기 위해 argument resolver 클래스 작성(파라미터로 현재 로그인 중인 유저의 정보를 담은 User 객체 전달)

* @CurrentUser (new)
- custom annotation으로써 해당 annotation이 붙어 있으면 argument resolver가controller로 파라미터를 전달할 때
설정에 따라 파라미터를 가공해서 전달

* WebConfig (new)
- argument resolver를 스프링에 등록하여 IoC 컨테이너가 이를 활성화 시킬
  수 있도록 함

* UserRepository/Mapper (update)
- password Update관련 Query, 메소드 추가

* UserService (update)
- 전달받은 user 객체와 userPasswordUpdateRequest 객체를 전달 받아
  userPasswordUpdate를 수행하는 로직 추가(first draft로 차후 수정 필수)

* 테스트 관련
- UserInfoUpdateTest (new)
: password update에 대한 비즈니스 로직을 테스트하는 단위 테스트
추가(성공/예외 케이스)

Commit:
[#4] 유저 비밀번호 변경기능 구현

* UserController (new)
- 유저 비밀번호 변경 요청을 보낼 `userPasswordUpdate` 메소드 작성

* UserService (update)
: `userPasswordUpdate` 메소드 refactoring (method extract)
- 현재 비밀번호를 비교하는 과정, 그리고 새로운 비밀번호가 일치하는지확인하는 과정에서 코드의 중복이 발생되어 리팩토링

* 테스트 관련(new)
1. UserControllerTest
- 유저 비밀번호 변경 요청 테스트 추가

2. UserRepositoryTest
- DB에서 비밀번호가 제대로 변경되는지 확인하기 위한 테스트 추가

f-lab-dev commented on src/main/java/me/soo/helloworld/util/CurrentUserArgumentResolver.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * 적용 시점
+ * - Client Request -> Dispatcher Servlet -> Request Handler 매핑 -> Handler Adapter 매핑
+ *   -> Interceptor 처리 -> ** Argument Resolver 처리 ** (이후 Message Converter 처리 -> Controller Method Invoke)
```
어떻게 이 클래스에 대해 알게 되었냐고 물어보려했는데 꼼꼼하게 잘 써주셨네요~ 👍

f-lab-dev commented on src/main/java/me/soo/helloworld/util/CurrentUserArgumentResolver.java:
```diff
@@ -0,0 +1,53 @@
+                                  NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
+
+        String currentUserId = loginService.getCurrentUserId();
+        User currentUser = userMapper.getUserById(currentUserId);
```
이렇게 하면 로그인한 유저의 정보를 받아올때 무조건 DB를 한번 갔다와야하네요.
`userId`만 필요할 수도 있는데 계속 유저를 조회하는거는 DB에 부담이 되지 않을까요?

msugo1 replied:
DB연산이 가장 부담이 큰 작업 중 하나라는 것을 잊어버리고 있었네요!

resolver에서는 `userId`만 받아와서 컨트롤러에 `String` 파라미터로 `userId` 주입시킨 후 서비스 계층에서 필요할 때만 DB에서 값을 가져오는 식으로 변경하고자 합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -17,27 +18,52 @@
+        userRepository.updateUserPassword(currentUser.getUserId(), encodedNewPassword);
+    }
+
+    private String encodePassword(String rawPassword) {
```
이거는 간단하게 `passwordEncoder.encode(rawPassword);`를 감싸는 역할만 하는데 이 메소드를 굳이 만들 필요가 있을까요?

msugo1 replied:
너무 간단한 메소드까지 메소드 추출을 활용하려고 했어서 그냥 제거하고 원래대로 변경하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -17,27 +18,52 @@
+        passwordMatchingCheck(userPasswordRequest.getCurrentPassword(), currentUser.getPassword());
+
+        String encodedNewPassword = passwordEncoder.encode(userPasswordRequest.getNewPassword());
+        // 새로운 비밀번호와 확인 값이 일치하지 않으면 예외발생
```
보통 비밀번호와 비밀번호확인은 사용자가 오타를 내는것을 방지하기 위한 수단으로 클라이언트에서 쓰입니다.

서버에서 이 로직을 적용할 필요가 있을까요?

msugo1 replied:
클라이언트 쪽에서 사용자 비밀번호 체크를 마무리 한 뒤 서버쪽으로 넘겨주는 것은 모르고 있었습니다.

그렇다면 인증은 더이상 서버 쪽의 몫이 아니기 때문에 체크하는 로직은 제거하였고 단순히 받은 비밀번호를 업데이트만 하는 것으로 변경하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -65,4 +67,17 @@
+                                                   @CurrentUser User user) {
+        try {
+            userService.userPasswordUpdate(userPasswordRequest, user);
+            loginService.logout();
```
비밀번호를 변경했을 때 다시 로그인을 시키려고 하시는 이유가 있으실까요?

msugo1 replied:
비밀번호가 변경되었으니 다시 로그인하게 해서 세션도 새로 등록해줄 필요가 있을 거라고 생각했는데 혹시 불필요한 과정일까요?

Commit:
[#4] 피드백 반영

* UserService (modify)
- `userPasswordUpdate` 메소드에 대한 불필요한 인증 절차 제거
  (클라이언트 쪽에서 비밀번호와 비밀번호 확인 절차가 이루어지므로
  서버에서는 이루어질 필요 없음)
- 불필요한 메소드 추출 제거(encodePassword와 isMatchingPassword 메소드)

* CurrentUserArgumentResolver (modify)
- User 객체 자체가 아닌 세션에 저장되어 있는 userId만 받아오도록 수정
  1) 매번 User 객체를 DB에서 받아오는 것은 비효율적(DB연산 자체가 큰
  부담이고, User 객체 전체가 아닌 userId만 필요한 경우도 있기 때문에)
  2) 따라서 User 객체 자체가 필요할 경우 Controller 계층에서 위의 resolver를
     통해 userId를 전달해주고 Service 계층에서 Repository 계층을 통해
     User 객체를 받아오는 것이 더 효율적일 것으로 판단

* 테스트 관련
- 불필요한 비밀번호 인증 로직을 제거했으므로 UserUpdate 테스트에서도 비밀번호 인증관련 로직
  제거

msugo1 committed:
Merge branch 'develop' into feature/4
